### PR TITLE
Cluster deploy

### LINKED ISSUE #95: K8SSAND-843 ⁃ #95 - k8ssandra-operator should support cluster-scoped deployments
Currently k8ssandra-operator is namespace-scoped. It only watches objects in specified namespaces which defaults to the namespace in which the operator is deployed. Interestingly the operator can be configured to watch different namespaces in remote clusters as compared to the ones it watches in the control plane cluster.

We should be able to configure the operator to be cluster-scoped such that it will watch object across all namespaces. Furthermore we should be able to create a multi-DC C* cluster (within a single k8s cluster) where each DC is in a different namespace.

There are a number of things that need (or may need) to be done for this including:

* Implement changes in operator code
  *  Some initial work has been started in #116. I did an initial review and pointed out some other changes that are needed.
  * K8ssandraClusterReconciler configures a watch on Stargate and CassandraDatacenter objects with a mapping function that checks for the cluster.  We may need to update the value of the label to also include the namespace of the K8ssandra cluster (or alternatively add a label for the namespace of the K8ssandraCluster).
*   Update kustomize manifests 
*  Update Helm chart
* Add `envtest` integration tests for
  * Multi-DC cluster in a single cluster across multiple namespaces
  * Multi-DC cluster in multiple cluster across multiple namespaces
* Add e2e tests for
  * Multi-DC cluster in a single cluster across multiple namespaces
  * Multi-DC cluster in multiple cluster across multiple namespaces 
* Update install guide 

It might make sense to break some of this up into multiple tickets.

Lastly note that the work describe here should also take care of #19.

┆Issue is synchronized with this [Jira Task](https://k8ssandra.atlassian.net/browse/K8SSAND-843) by [Unito](https://www.unito.io)
┆Reviewer: 5e66ca0cae59460cfb01a13c
┆Issue Number: K8SSAND-843
┆Priority: Medium

### PR DESCRIPTION
**Pull request by jsanda**
**What this PR does**:

**Which issue(s) this PR fixes**:
Fixes #95

**Checklist**
- [ ] Changes manually tested
- [ ] Automated Tests added/updated
- [ ] Documentation added/updated
- [ ] CHANGELOG.md updated (not required for documentation PRs)
- [ ] CLA Signed:  [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

File: controllers/replication/secret_controller_test.go [hunk 1]
```diff
@@ -253,6 +265,8 @@ func generateSecrets(namespace string) []*corev1.Secret {
 				Name:      "test-secret-first",
 				Labels: map[string]string{
 					"secret-controller": "test",
+					"cluster-namespace": namespace,
+					"cluster-name":      "k8ssandra",
```
Comment (adutra) [inline controllers/replication/secret_controller_test.go:269]: Isn't it better to use the available constants here?

File: pkg/secret/replicated.go [hunk 2]
```diff
@@ -201,13 +194,14 @@ func requiresUpdate(current, desired *replicationapi.ReplicatedSecret) bool {
 	return false
 }
 
-func getManagedObjectMeta(name, namespace, clusterName string) metav1.ObjectMeta {
+func getManagedObjectMeta(name string, kcKey client.ObjectKey) metav1.ObjectMeta {
```
Comment (adutra) [inline pkg/secret/replicated.go:197]: This will hopefully remove the risk of confusing `kc.Name` with `kc.Spec.Cassandra.Cluster` 👍

**Review by adutra — APPROVED** [review]
LGTM (if this is not in WIP state anymore of course).

File: controllers/replication/secret_controller_test.go (hunk 1 above)
Comment (jsanda) [reply controllers/replication/secret_controller_test.go:269]: Yeah, I'll update it.

**Comment by jsanda** [comment]
I still have some outstanding tasks:

* Add e2e test that uses cluster-scope deployment
    * I think a single test is sufficient for the PR. More test scenarios can be added in follow up PRs
* Update Makefile targets
* Do another review of code to see if there any other queries where namespace predicate should be removed

File: test/e2e/suite_test.go [hunk 3]
```diff
@@ -235,53 +328,55 @@ func applyPollingDefaults() {
...
 
-func cleanUp(t *testing.T, namespace string, f *framework.E2eFramework, deployTraefik bool) error {
+func cleanUp(t *testing.T, f *framework.E2eFramework, opts *e2eTestOpts) error {
 
-	if t.Failed() {
-		f.DumpOperatorLogs(namespace)
-		f.DumpCassandraLogs(namespace)
-		f.DumpStargateLogs(namespace)
+	namespaces := make([]string, 0)
+	namespaces = append(namespaces, opts.operatorNamespace)
```
Comment (adutra) [inline test/e2e/suite_test.go:338]:
```suggestion
	namespaces = append(namespaces, opts.operatorNamespace)
	namespaces = append(namespaces, opts.sutNamespace)
```

Comment (adutra) [reply test/e2e/suite_test.go:338]: (I noticed that `sutNamespace` isn't getting deleted.)

File: Makefile [hunk 4]
```diff
@@ -158,10 +160,18 @@ single-reload: build manifests kustomize docker-build kind-load-image cert-manag
 multi-up: cleanup build manifests kustomize docker-build create-kind-multicluster kind-load-image-multi cert-manager-multi
 ## install the control plane
 	kubectl config use-context kind-k8ssandra-0
+ifeq ($(CLUSTER_SCOPE),true)
+	$(KUSTOMIZE) build config/deployments/control-plane-cluster-scope | kubectl apply -f -
```
[the remark this answers is not in the document]
Comment (jsanda) [reply Makefile:164]: It probably is. I missed it.

File: Makefile [hunk 5]
```diff
@@ -158,10 +160,18 @@ single-reload: build manifests kustomize docker-build kind-load-image cert-manag
 multi-up: cleanup build manifests kustomize docker-build create-kind-multicluster kind-load-image-multi cert-manager-multi
```
[the remark this answers is not in the document]
Comment (jsanda) [reply Makefile:160]: Just an oversight. I'll take another pass through Makefile. There might be other changes needed as well.

File: config/deployments/data-plane-cluster-scope/kustomization.yaml [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+kind: Kustomization
+
+namespace: k8ssandra-operator
+
+resources:
+  - ../../cass-operator-cluster-scope
+  - ../../default
+
+components:
+  - ../../components/clusterscope
```
[the remark this answers is not in the document]
Comment (jsanda) [reply config/deployments/data-plane-cluster-scope/kustomization.yaml:11]: I'll change it. The inconsistency will drive me crazy :)

File: pkg/secret/replicated.go [hunk 7]
```diff
@@ -52,14 +52,13 @@ func DefaultSuperuserSecretName(clusterName string) string {
 	return secretName
 }
 
-// ReconcileSecret creates a new secret with proper "managed-by" annotations, or ensure the existing secret has such
-// annotations.
-func ReconcileSecret(ctx context.Context, c client.Client, secretName, clusterName, namespace string) error {
+// ReconcileSecret creates the superUserSecret with proper annotations
```
[the remark this answers is not in the document]
Comment (jsanda) [reply pkg/secret/replicated.go:55]: I'll update accordingly.

File: pkg/secret/replicated.go [hunk 8]
```diff
@@ -158,19 +156,16 @@ func HasReplicatedSecrets(ctx context.Context, c client.Client, clusterName, nam
...
-
+func generateReplicatedSecret(kcKey client.ObjectKey, replicationTargets []replicationapi.ReplicationTarget) *replicationapi.ReplicatedSecret {
 	return &replicationapi.ReplicatedSecret{
-		ObjectMeta: getManagedObjectMeta(clusterName, namespace, clusterName),
+		ObjectMeta: getManagedObjectMeta(kcKey.Name, kcKey),
 		Spec: replicationapi.ReplicatedSecretSpec{
 			Selector: &metav1.LabelSelector{
-				MatchLabels: utils.ManagedByLabels(clusterName),
+				MatchLabels: map[string]string{
+					api.ManagedByLabel:                 api.NameLabelValue,
```
[the remark this answers is not in the document]
Comment (jsanda) [reply pkg/secret/replicated.go:165]: Nothing :( Looks I have more clean up to do from a messy rebase.

File: config/deployments/data-plane-cluster-scope/kustomization.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (adutra) [reply config/deployments/data-plane-cluster-scope/kustomization.yaml:11]: Glad to see that we share the same OCD symptoms :-D

**Comment by adutra** [comment]
FYI when running the cluster-scope e2e test locally, I noticed that the DCs are not getting deleted properly:

```
INFO[1009] deleting all K8ssandraClusters                Namespace=k8ssandra-operator
INFO[1009] deleting all Stargates                        Namespace=k8ssandra-operator
INFO[1019] deleting all Reapers                          Namespace=k8ssandra-operator
INFO[1029] deleting all CassandraDatacenters             Namespace=k8ssandra-operator
INFO[1039] deleting all ReplicatedSecrets                Namespace=k8ssandra-operator
INFO[1049] deleting namespace                            Context=kind-k8ssandra-1 Namespace=k8ssandra-operator
INFO[1049] deleting namespace                            Context=kind-k8ssandra-0 Namespace=k8ssandra-operator
INFO[1069] deleting all K8ssandraClusters                Namespace=test-0
INFO[1069] deleting all Stargates                        Namespace=test-0
INFO[1079] deleting all Reapers                          Namespace=test-0
INFO[1089] deleting all CassandraDatacenters             Namespace=test-0
INFO[1099] deleting all ReplicatedSecrets                Namespace=test-0
    suite_test.go:375: failed to delete replicated secrets: timed out waiting for the condition
INFO[1279] deleting namespace                            Context=kind-k8ssandra-1 Namespace=test-0
INFO[1279] deleting namespace                            Context=kind-k8ssandra-0 Namespace=test-0
    suite_test.go:379: failed to delete namespace: timed out waiting for the condition
INFO[1459] deleting all K8ssandraClusters                Namespace=test-1
INFO[1459] deleting all Stargates                        Namespace=test-1
INFO[1469] deleting all Reapers                          Namespace=test-1
INFO[1479] deleting all CassandraDatacenters             Namespace=test-1
    suite_test.go:371: failed to delete datacenters: timed out waiting for the condition
INFO[1659] deleting all ReplicatedSecrets                Namespace=test-1
INFO[1669] deleting namespace                            Context=kind-k8ssandra-1 Namespace=test-1
INFO[1669] deleting namespace                            Context=kind-k8ssandra-0 Namespace=test-1
    suite_test.go:379: failed to delete namespace: timed out waiting for the condition
INFO[1849] deleting all K8ssandraClusters                Namespace=test-2
INFO[1849] deleting all Stargates                        Namespace=test-2
INFO[1859] deleting all Reapers                          Namespace=test-2
INFO[1869] deleting all CassandraDatacenters             Namespace=test-2
    suite_test.go:371: failed to delete datacenters: timed out waiting for the condition
INFO[2049] deleting all ReplicatedSecrets                Namespace=test-2
INFO[2059] deleting namespace                            Context=kind-k8ssandra-1 Namespace=test-2
INFO[2059] deleting namespace                            Context=kind-k8ssandra-0 Namespace=test-2
    suite_test.go:379: failed to delete namespace: timed out waiting for the condition
``
`

If that happens again I'll have a closer look.

File: pkg/secret/replicated.go [hunk 9]
```diff
@@ -208,8 +203,14 @@ func requiresUpdate(current, desired *replicationapi.ReplicatedSecret) bool {
...
-func getManagedObjectMeta(name, namespace, clusterName string) metav1.ObjectMeta {
-	meta := metav1.ObjectMeta{Name: name, Namespace: namespace}
-	utils.SetManagedBy(&meta, clusterName)
-	return meta
+func getManagedObjectMeta(name string, kcKey client.ObjectKey) metav1.ObjectMeta {
+	return metav1.ObjectMeta{
+		Name:      name,
+		Namespace: kcKey.Namespace,
+		Labels: map[string]string{
+			api.ManagedByLabel:                 api.NameLabelValue,
```
[the remark this answers is not in the document]
Comment (jsanda) [reply pkg/secret/replicated.go:211]: Is `utils.SetManagedBy` redundant with `utils.ManagedByLabels`?

**Comment by jsanda** [comment]
>FYI when running the cluster-scope e2e test locally, I noticed that the DCs are not getting deleted properly

I noticed that same behavior. Everything should be getting deleted with the finalizer. I did some more testing locally to confirm that 1) the deletions do happen with the finalizer and 2) the timeouts are coming from the `E2eFramework.deleteAllResources` method.

**Comment by jsanda** [comment]
The problem with the clean up function(s) is that it wasn't waiting for the K8ssandraCluster deletion to finish. It would immediately move on to delete the operator and namespaces which left things in a Terminating state indefinitely. It should be fixed now 🤞.

File: test/e2e/suite_test.go [hunk 10]
```diff
@@ -55,13 +55,54 @@ func TestOperator(t *testing.T) {
...
+		fixture:  "multi-dc",
+	}))
+	t.Run("CheckStargateApisWithMultiDcCluster", e2eTest(ctx, &e2eTestOpts{
+		testFunc:      checkStargateApisWithMultiDcCluster,
+		fixture:       "multi-dc-stargate",
+		deployTraefik: true,
+	}))
+	t.Run("CreateSingleReaper", e2eTest(ctx, &e2eTestOpts{
+		testFunc: createSingleReaper,
+		fixture:  "single-dc-reaper",
```
Comment (adutra) [inline test/e2e/suite_test.go:90]: When #223 is merged, you will need `deployTraefik: true` for both Reaper tests.

**Review by adutra — APPROVED** [review]

**Pull request merged**